crates/turborepo-filewatch/src/lib.rs

```diff
@@ -0,0 +1,793 @@
+    while let Some(parent) = current.parent() {
+        current = parent;
+        let watch_result = watcher.watch(current.as_std_path(), RecursiveMode::NonRecursive);
+        if is_permission_denied(&watch_result) {
+            // It is expected we hit a permission denied error at some point. We won't
+            // get notifications if someone e.g. deletes all of /home
+            break;
+        } else {
+            watch_result?;
+        }
+    }
+    Ok(())
+}
+
+#[cfg(feature = "watch_recursively")]
+fn watch_recursively(
+    root: &Path,
+    watcher: &mut Backend,
+    sender: Option<(
+        Instant,
+        &broadcast::Sender<Result<DebouncedEvent, Vec<NotifyError>>>,
+    )>,
+) -> Result<(), WatchError> {
+    for dir in WalkDir::new(root).follow_links(false).into_iter() {
+        let dir = dir?;
+        if dir.file_type().is_dir() {
+            watcher.watch(dir.path(), RecursiveMode::NonRecursive)?;
+        }
+        if let Some((instant, sender)) = sender.as_ref() {
+            let create_kind = if dir.file_type().is_dir() {
+                CreateKind::Folder
+            } else {
+                CreateKind::File
+            };
+            let event = DebouncedEvent {
+                event: Event {
+                    paths: vec![dir.path().to_owned()],
+                    kind: EventKind::Create(create_kind),
+                    attrs: EventAttributes::default(),
+                },
+                time: *instant,
+            };
+            // It's ok if we fail to send, it means we're shutting down
+            let _ = sender.send(Ok(event));
+        }
+    }
+    Ok(())
+}
+
+fn run_watcher(
+    root: &AbsoluteSystemPath,
+    sender: mpsc::Sender<DebounceEventResult>,
+) -> Result<Debouncer<Backend, FileIdMap>, WatchError> {
+    #[cfg(target_os = "macos")]
+    let mut debouncer = make_debouncer(move |res| {
+        let _ = sender.blocking_send(res);
+    })?;
+    #[cfg(not(target_os = "macos"))]
+    let mut debouncer = new_debouncer(Duration::from_millis(1), None, move |res| {
+        let _ = sender.blocking_send(res);
+    })?;
+
+    #[cfg(any(target_os = "macos", target_os = "windows"))]
+    debouncer
+        .watcher()
+        .watch(&root.as_std_path(), RecursiveMode::Recursive)?;
+    #[cfg(feature = "watch_recursively")]
+    {
+        // Don't synthesize initial events
+        watch_recursively(root.as_std_path(), debouncer.watcher(), None)?;
+    }
+    #[cfg(feature = "watch_ancestors")]
+    watch_parents(root, debouncer.watcher())?;
+    Ok(debouncer)
+}
+
+#[cfg(target_os = "macos")]
+fn make_debouncer<F: DebounceEventHandler>(
+    event_handler: F,
+) -> Result<Debouncer<Backend, FileIdMap>, notify::Error> {
+    new_debouncer_opt::<F, FsEventWatcher, FileIdMap>(
+        Duration::from_millis(1),
+        None,
+        event_handler,
+        FileIdMap::new(),
+        notify::Config::default(),
+    )
+}
+
+//#[cfg(target_os = "macos")]
+async fn wait_for_cookie(
```

Comment:
Could we add some documentation on what exactly the cookie is here?

Reply:
Yup, done.